### ISSUE TITLE
pitchfork build err :: 

### ISSUE THREAD

**Issue by EricDeveaud** [PacificBiosciences/blasr#332]
hello, 

trying to compile and install blasr using pitchfork, cause I need bam support  
and saw issue #253

I followed the given documentation form pitchfork readme.
```
git clone https://github.com/PacificBiosciences/pitchfork.git pitchfork
cd pitchfork
make init PREFIX=/tmp/mybuild            
make blasr PREFIX=/tmp/mybuild
```

it fails on pbbam with following lasts lines 

```
[ 76%] Building CXX object src/CMakeFiles/pbbam.dir/ZmwReadStitcher.cpp.o
[ 78%] Building CXX object src/CMakeFiles/pbbam.dir/ZmwQuery.cpp.o
[ 79%] Building CXX object src/CMakeFiles/pbbam.dir/ZmwTypeMap.cpp.o
[ 80%] Building CXX object src/CMakeFiles/pbbam.dir/pugixml/pugixml.cpp.o
[ 81%] Linking CXX shared library ../lib/libpbbam.so
/usr/bin/ld: /tmp/mybuild/src/htslib/lib/libhts.a(bgzf.o): relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
/tmp/mybuild/src/htslib/lib/libhts.a: could not read symbols: Bad value
collect2: error: ld returned 1 exit status
make[4]: *** [lib/libpbbam.so] Error 1
make[4]: Leaving directory `/tmp/pitchfork/workspace/pbbam/build'
make[3]: *** [src/CMakeFiles/pbbam.dir/all] Error 2
make[3]: Leaving directory `/tmp/pitchfork/workspace/pbbam/build'
make[2]: *** [all] Error 2
make[2]: Leaving directory `/tmp/pitchfork/workspace/pbbam/build'
make[1]: *** [do-build] Error 2
make[1]: Leaving directory `/tmp/pitchfork/ports/pacbio/pbbam'
make: *** [pbbam] Error 2
```

best regards

Eric

edit markdown typo

**Comment by EricDeveaud** [comment]
additional info.

ppbam build command issued by pitchfork

```
git clone git://github.com/PacificBiosciences/pbbam /tmp/pitchfork/workspace/pbbam
Cloning into '/tmp/pitchfork/workspace/pbbam'...
remote: Counting objects: 4403, done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 4403 (delta 0), reused 0 (delta 0), pack-reused 4398
Receiving objects: 100% (4403/4403), 14.42 MiB | 5.00 MiB/s, done.
Resolving deltas: 100% (3204/3204), done.
Checking connectivity... done.
cd /tmp/pitchfork/workspace/pbbam && git checkout HEAD
Your branch is up-to-date with 'origin/master'.
mkdir -p /tmp/pitchfork/workspace/pbbam/build
cd /tmp/pitchfork/workspace/pbbam/build \
     && cmake \
            -DPacBioBAM_build_shared=ON \
            -DPacBioBAM_build_docs=OFF \
            -DPacBioBAM_build_tests=OFF \
            -DHTSLIB_INCLUDE_DIRS=/tmp/mybuild/src/htslib/include \
            -DHTSLIB_LIBRARIES=/tmp/mybuild/src/htslib/lib/libhts.a \
            -DZLIB_INCLUDE_DIRS=/tmp/mybuild/include \
            -DZLIB_LIBRARIES=/tmp/mybuild/lib/libz.so \
            -DBoost_INCLUDE_DIRS=/tmp/mybuild/include \
            -DCMAKE_BUILD_TYPE=RelWithDebInfo \
            -DCMAKE_SKIP_BUILD_RPATH=FALSE \
```

**Comment by mjhsieh** [comment]
I will need to know some of the information first, such as the os distribution, which compiler used and usage of gnu env modules. Thanks. first glance is that libhts.a might have includes some wrong .a file from the system.

**Comment by EricDeveaud** [comment]
os is:
```
bigmess:/tmp/pitchfork > cat /etc/centos-release 
CentOS release 6.8 (Final)
```

we use module environment: http://modules.sourceforge.net/ 

loaded module for the pitchfork blasr buidl are

```
bigmess:/tmp/pitchfork > module list 
Currently Loaded Modulefiles:
  1) gcc/4.9.0     2) git/2.2.2     3) cmake/3.2.1
```

**Comment by mhsieh** [comment]
Can you tell me the $CFLAGS and $LDFLAGS of your current shell?

**Comment by EricDeveaud** [comment]
sorry for the long delay. 
hereare the requested infos
```
bigmess:/tmp/pitchfork > echo $CFLAGS                             
-Wall -g
bigmess:/tmp/pitchfork > echo $LDFLAGS
-L/local/gensoft2/exe/git/2.2.2/lib -L/local/gensoft2/exe/gcc/4.9.0/lib64 -L/local/gensoft2/exe/gcc/4.9.0/lib
```
NB just tried with the las pitchfork revision, and got same errormessage.

process was:
```
bigmess:/tmp > rm -rf pitchfork mybuild 
bigmess:/tmp > git clone https://github.com/PacificBiosciences/pitchfork.git
bigmess:/tmp > cd pitchfork 
bigmess:/tmp/pitchfork > make init PREFIX=/tmp/mybuild            
bigmess:/tmp/pitchfork > make blasr PREFIX=/tmp/mybuild | tee log                     
```

and error was again:

```
[ 81%] Linking CXX shared library ../lib/libpbbam.so
/usr/bin/ld: /tmp/mybuild/src/htslib/lib/libhts.a(bgzf.o): relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
/tmp/mybuild/src/htslib/lib/libhts.a: could not read symbols: Bad value
collect2: error: ld returned 1 exit status
make[4]: *** [lib/libpbbam.so] Error 1
```

regards

Eric

**Comment by mjhsieh** [comment]
I suspect that your git module include an in compatible zlib, please git pull --rebase under pitchfork directory, do a `make _startover PREFIX=/tmp/mybuild` and redo your build a gain.

**Comment by EricDeveaud** [comment]
sorry for the long delay. 
hereare the requested infos
```
bigmess:/tmp/pitchfork > echo $CFLAGS                             
-Wall -g
bigmess:/tmp/pitchfork > echo $LDFLAGS
-L/local/gensoft2/exe/git/2.2.2/lib -L/local/gensoft2/exe/gcc/4.9.0/lib64 -L/local/gensoft2/exe/gcc/4.9.0/lib
```
NB just tried with the las pitchfork revision, and got same errormessage.

process was:
```
bigmess:/tmp > rm -rf pitchfork mybuild 
bigmess:/tmp > git clone https://github.com/PacificBiosciences/pitchfork.git
bigmess:/tmp > cd pitchfork 
bigmess:/tmp/pitchfork > make init PREFIX=/tmp/mybuild            
bigmess:/tmp/pitchfork > make blasr PREFIX=/tmp/mybuild | tee log                     
```

and error was again:

```
[ 81%] Linking CXX shared library ../lib/libpbbam.so
/usr/bin/ld: /tmp/mybuild/src/htslib/lib/libhts.a(bgzf.o): relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
/tmp/mybuild/src/htslib/lib/libhts.a: could not read symbols: Bad value
collect2: error: ld returned 1 exit status
make[4]: *** [lib/libpbbam.so] Error 1
```

regards

Eric

**Comment by EricDeveaud** [comment]
@mjhsieh thanks for your help and patience

done what you suggested. see

```
bigmess:/tmp/pitchfork > load git
bigmess:/tmp/pitchfork > git pull --rebase
Current branch master is up to date.
bigmess:/tmp/pitchfork > make _startover PREFIX=/tmp/mybuild
rm -f mk//../initialized.o
This will erase everything in /tmp/mybuild/, staging/ and /tmp/pitchfork/workspace/ directories.
Are you sure? y
+ rm -rf '/tmp/mybuild/*' /tmp/mybuild/.Python staging/boost-1.60.0 staging/cmake-3.4.1 staging/hdf5-1.8.16 staging/ncurses-6.0 staging/samtools-1.3.1 staging/zlib-1.2.8 /tmp/pitchfork/workspace/boost-1.60.0 /tmp/pitchfork/workspace/cmake-3.4.1 /tmp/pitchfork/workspace/gtest-1.7.0 /tmp/pitchfork/workspace/hdf5-1.8.16 /tmp/pitchfork/workspace/htslib /tmp/pitchfork/workspace/ncurses-6.0 /tmp/pitchfork/workspace/pbbam /tmp/pitchfork/workspace/samtools-1.3.1 /tmp/pitchfork/workspace/zlib-1.2.8 ports/pacbio/htslib/build.log ports/thirdparty/boost/build.log ports/thirdparty/boost/config.log ports/thirdparty/cmake/build.log ports/thirdparty/cmake/config.log ports/thirdparty/cmake/install.log ports/thirdparty/hdf5/build.log ports/thirdparty/hdf5/config.log ports/thirdparty/hdf5/install.log ports/thirdparty/ncurses/build.log ports/thirdparty/ncurses/config.log ports/thirdparty/ncurses/install.log ports/thirdparty/samtools/build.log ports/thirdparty/samtools/config.log ports/thirdparty/samtools/install.log ports/thirdparty/zlib/build.log ports/thirdparty/zlib/config.log ports/thirdparty/zlib/install.log
+ rm -f 'ports/*/*/*_cram.xml' 'ports/*/*/*_nose.xml'
+ test -d /tmp/pitchfork/.git/pip/wheels
+ true
bigmess:/tmp/pitchfork > module unload git 
bigmess:/tmp/pitchfork > module load gcc/4.9.0 cmake/3.2.1
bigmess:/tmp/pitchfork > make blasr
```

and same problem.

```
[ 81%] Linking CXX shared library ../lib/libpbbam.so
/usr/bin/ld: /tmp/pitchfork/deployment/src/htslib/lib/libhts.a(bgzf.o): relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
/tmp/pitchfork/deployment/src/htslib/lib/libhts.a: could not read symbols: Bad value
collect2: error: ld returned 1 exit status
make[4]: *** [lib/libpbbam.so] Error 1
```

what amaze me is that I am abble to compile pbbam from https://github.com/PacificBiosciences/pbbam.git
in the same environnement.using thefollowing

```
module purge && module load  \
                gcc/4.9.0 \
                cmake/3.2.1 \
                boost/1_61_0 \
                zlib/1.2.8 && \
                cmake \
                        -DCMAKE_INSTALL_PREFIX=/local/gensoft2/lib/pbbam/0.7.4 \
                        -DPacBioBAM_build_shared=ON \
                        -DBOOST_ROOT=$BOOST_ROOT \
                        -DPacBioBAM_build_tests=OFF \
                        -DCMAKE_SKIP_INSTALL_RPATH=ON \
                        -DCMAKE_SKIP_RPATH=ON \
                        /local/gensoft2/src/pbbam/pbbam-0.7.4
```

**Comment by mhsieh** [comment]
I am now very intrigued by this libhts issue, can you send me (mhsieh at pacb.com) the log under ports/pacbio/htslib ? I suspect that the ingenuity of your environment modules is defeating pitchfork's control-freak design.

**Comment by mhsieh** [comment]
In the meantime, you can try writing a settings.mk file under pitchfork folder
```
PREFIX=/tmp/mybuild
HAVE_ZLIB=/local/gensoft2/exe/zlib/1.2.8
HAVE_CMAKE=/local/gensoft2/exe/cmake/3.2.1/bin/cmake
HAVE_BOOST=/local/gensoft2/exe/boost/1_61_0
```

then do:
```shell
$ module load gcc/4.9.0 git/2.2.2
$ echo y | make _startover
$ make pbbam
```

**Comment by mhsieh** [comment]
Sorry I forgot to merge the PR, can you try again?

**Comment by EricDeveaud** [comment]
@mjhsieh build log sent.
same problem after git update 

```
bigmess:/tmp/pitchfork > git log -n  2
commit 172d78b21a72f1044a10c1f2f04152516fef9dbf
Author: Mengjuei Hsieh <mhsieh@pacificbiosciences.com>
Date:   Tue Feb 21 21:55:00 2017 -0800

    hotfix - update htslib version

commit 97222268aa1bf9350364f9fff86d8e9b9ad4c472
Author: Mengjuei Hsieh <mjhsieh@users.noreply.github.com>
Date:   Tue Feb 21 19:03:20 2017 -0800

    UGLY FIX - clash with module author (#379)
    
    https://github.com/PacificBiosciences/blasr/issues/332
```

and make blasr ends with the same pbbam problem.

currently running make pbbam with settings.mk set

**Comment by EricDeveaud** [comment]
make blasr success with the combination of last version and settings.mk

thanks for your help 

Eric
